Parallel implementation

Vegoo89 opened the pull request:
No modification in standard module so it can still be used the same way as before. Example imports:

`import sparse_dot_topn.sparse_dot_topn as ct_original`
`import sparse_dot_topn.sparse_dot_topn_threaded as ct_thread`

Example function calls:

`ct_original.sparse_dot_topn(row_len, col_len, train_ptr, train_indices, train_data, g_ptr, g_indices, g_data, best_entries_len, probability_above, c_ptr, c_indicies, c_data)`

`ct_thread.sparse_dot_topn_threaded(row_len, col_len, train_ptr, train_indices, train_data, g_ptr, g_indices, g_data, best_entries_len, probability_above, c_ptr, c_indicies, c_data, n_jobs)`

Speed comparison using two matrices:
* `tf_idf_train_names` is `<100000x61134 sparse matrix of type '<class 'numpy.float64'>'`
* `tf_idf_g_names` is `<450256x61134 sparse matrix of type '<class 'numpy.float64'>'`

Results using original function and threaded function with 1 thread:

![cosine_topn_comparision](https://user-images.githubusercontent.com/34243210/61778340-dc402300-adfe-11e9-8d4d-efd568c2da4b.PNG)

Threaded function results using 8 threads:

![cosine_topn_8_threads](https://user-images.githubusercontent.com/34243210/61778641-69837780-adff-11e9-85c6-80e404b5cf5d.PNG)

Only common file with major changes is `setup.py`

It is now building two extensions and pack them into single python module. Also I modified `extra_compile_args` and used performance optimization instead of size. 

Tested it on Linux / Windows - don't know how/if it will compile on Mac with these settings.

Also it's worth to mention that threading speeds up process significally only under certain conditions. If there is cpu -> memory bottleneck (like on my work laptop) it will slow down calculations if you run more than 1 thread.

Commit:
Commit squash #2

Commit squash

First Commit

Parallel implementation

Comparision added
Awesome cossim top modified
Duplicated functions/structs moved to single file

Readme and example update

aerdem4 commented on sparse_dot_topn/sparse_dot_topn_parallel.cpp:
```diff
@@ -0,0 +1,198 @@
+//
+static bool candidate_cmp(candidate c_i, candidate c_j) { return (c_i.value > c_j.value); }
+
+void inner_sparse_function(int start_row, int end_row, int n_col_inner,
```
This seems like code duplication from the original function. Can you call the original function instead of copying the functionality?

aerdem4 commented:
@Vegoo89 Thanks for your PR! Can you add your test to comparison.py? Can you also squash the commits into one?

@ymwdalex can you review too?

Vegoo89 commented:
> @Vegoo89 Thanks for your PR! Can you add your test to comparison.py? Can you also squash the commits into one?
> 
> @ymwdalex can you review too?

Updated comparison files and squashed commits

ymwdalex commented:
@Vegoo89 thanks for the pull request. It looks good!

Could you please also update readme.md file to show this new parallel implemention?

Vegoo89 commented:
> @Vegoo89 thanks for the pull request. It looks good!
> 
> Could you please also update readme.md file to show this new parallel implemention?

Done
